### ISSUE TITLE
[v5] Holding type invariants during Map and Traverse

### ISSUE THREAD

**Issue by NonExisto** [louthy/language-ext#1420]
In scope of collection type some of those have certain invariants aka type definition and purpose laws. To be concrete let check out HashSet<T> which promises to hold only unique value. Now lets modify existing unit test to check this invariant:
```c#
[Fact]
public void HashSetArrCrossProductTypeInvariant()
{
    var ma = HashSet(Array(1, 1, 1));

    var mb = ma.KindT<HashSet, Arr, Arr<int>, int>()
               .SequenceM()
               .AsT<Arr, HashSet, HashSet<int>, int>()
               .As();
           

    var mc = Array(HashSet(1));

    Assert.True(mb == mc); // Fails due to mb has 3 elements
}
```

I see same issue with Set<T>. And its Map function would gladly create a set of same integer value from complete valid collection:
```c#
[Fact]
public void SetTypeInvariantInMap()
{
    var ma = Set(1,2,3,4);
    var mb = ma.Map(x => 1);

    var mc = Set(1);

    Assert.True(mb == mc); // Fails due to mb has 3 elements
}
```
Both unit tests are failing and that aren't looks correct to me.

**Comment by louthy** [comment]
Please report single issue reports.  The first one isn't a bug, it is expected behaviour.  The second one is an issue and will be fixed asap.

**Comment by louthy** [comment]
Fix in `v5.0.0-beta-39`
